compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorLowering.cpp

```diff
@@ -142,18 +144,58 @@ struct FoldI1SelectToBroadcast final : OpRewritePattern<arith::SelectOp> {
   }
 };
 
+struct SetMulAddFMF final : OpRewritePattern<arith::AddFOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(arith::AddFOp addOp,
+                                PatternRewriter &rewriter) const override {
+    auto mulOp = addOp.getLhs().getDefiningOp<arith::MulFOp>();
+    if (!mulOp) {
+      mulOp = addOp.getRhs().getDefiningOp<arith::MulFOp>();
+    }
+
+    if (!mulOp) {
+      return failure();
+    }
+
+    constexpr auto contract = arith::FastMathFlags::contract;
+    if (mulOp.getFastmath() != arith::FastMathFlags::none &&
+        addOp.getFastmath() != arith::FastMathFlags::none) {
+      return failure();
+    }
+
+    rewriter.replaceOpWithNewOp<arith::MulFOp>(
+        mulOp, mulOp.getType(), mulOp.getLhs(), mulOp.getRhs(), contract);
+    rewriter.replaceOpWithNewOp<arith::AddFOp>(
+        addOp, addOp.getType(), addOp.getLhs(), addOp.getRhs(), contract);
+
+    return success();
+  }
+};
+
 struct LLVMGPUVectorLoweringPass final
     : impl::LLVMGPUVectorLoweringPassBase<LLVMGPUVectorLoweringPass> {
   void getDependentDialects(DialectRegistry &registry) const override {
     registry.insert<affine::AffineDialect>();
     registry.insert<memref::MemRefDialect>();
     registry.insert<vector::VectorDialect>();
     registry.insert<scf::SCFDialect>();
+    registry.insert<math::MathDialect>();
   }
   void runOnOperation() override {
     auto funcOp = getOperation();
     MLIRContext *ctx = &getContext();
 
+    // Uplift arith ops to math.fma before lowering high level vector operation.
+    {
+      RewritePatternSet fmaPatterns(ctx);
+      fmaPatterns.add<SetMulAddFMF>(ctx);
```

Comment:
Maybe add this with a higher benefit so that it always runs before the uplift pattern?